src/ContentValidationListener.php

```diff
@@ -213,7 +215,35 @@ public function onRoute(MvcEvent $e)
         }
 
         $inputFilter->setData($data);
+
         if ($inputFilter->isValid()) {
+            if (isset($this->config[$controllerService]['use_raw_data'])
+                    && $this->config[$controllerService]['use_raw_data'] === true) {
+                $dataContainer->setBodyParams($data);
+
+                return;
+            }
+
+            if (!$inputFilter instanceof BaseInputFilter) {
+                return;
+            }
+
+            $bodyParams = $inputFilter->getValues();
+            $unknown = $inputFilter->getUnknown();
+
+            if (count($unknown) !== 0) {
+                if ($this->isAllowsOnlyFieldsInFilter($controllerService)) {
+                    $fields = implode(', ', array_keys($unknown));
+                    $detail = sprintf('Unrecognized fields: %s', $fields);
+                    $problem = new ApiProblem(Response::STATUS_CODE_422, $detail);
+
+                    return new ApiProblemResponse($problem);
+                } elseif (count($inputFilter->getInputs()) === 0) {
+                    $bodyParams = $unknown;
+                }
+            }
+
+            $dataContainer->setBodyParams($bodyParams);
```

Comment:
If `allow_only_fields_in_filter` is boolean false and/or not defined, I'd expect the unknown values to be _merged_ with the data returned from the input filter. I'll make that change on merge as well.